ANALYZING MOBILE APP DATA

Combining skills learned to perform practical data analysis

In [1]:
from csv import reader

In [44]:
apple = list(reader(open('AppleStore.csv')))
google = list(reader(open('googleplaystore.csv')))

In [45]:
def explore_data(data_set,start,end,rows_and_columns=False):
    data_set_slice = data_set[start:end]
    for row in data_set_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row
    if rows_and_columns:
        print('Number of rows:', len(data_set))
        print('Number of columns:', len(data_set[0]))        

In [47]:
explore_data(apple,0,2,True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


Number of rows: 7198
Number of columns: 16


In [48]:
explore_data(google,0,5)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']




In [49]:
print(google[0])
print(google[10472])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Xposed Wi-Fi-Pwd', 'PERSONALIZATION', '3.5', '1042', '404k', '100,000+', 'Free', '0', 'Everyone', 'Personalization', 'August 5, 2014', '3.0.0', '4.0.3 and up']


REMOVING DUPLICATES

The code below is used to search for duplicate apps.

In [51]:
unique_apps = []
duplicates = []

for data in google[1:]: #loop through google play store apps
    app = data[0]# select the app name
    
    if app not in unique_apps:
        unique_apps.append(app)
    else:
        duplicates.append(data)

In [52]:
print(unique_apps)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'Coloring book moana', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint', 'Pixel Draw - Number Art Coloring Book', 'Paper flowers instructions', 'Smoke Effect Photo Maker - Smoke Editor', 'Infinite Painter', 'Garden Coloring Book', 'Kids Paint Free - Drawing Fun', 'Text on Photo - Fonteee', 'Name Art Photo Editor - Focus n Filters', 'Tattoo Name On My Photo Editor', 'Mandala Coloring Book', '3D Color Pixel by Number - Sandbox Art Coloring', 'Learn To Draw Kawaii Characters', 'Photo Designer - Write your name with shapes', '350 Diy Room Decor Ideas', 'FlipaClip - Cartoon animation', 'ibis Paint X', 'Logo Maker - Small Business', "Boys Photo Editor - Six Pack & Men's Suit", 'Superheroes Wallpapers | 4K Backgrounds', 'Mcqueen Coloring pages', 'HD Mickey Minnie Wallpapers', 'Harley Quinn wallpapers HD', 'Colorfit - Drawing & Coloring', 'Animated Photo Editor', 'Pencil Sketch Drawing', 'Easy Realistic Drawing Tut

In [54]:
print(duplicates[0:2])

[['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up'], ['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']]


THERE ARE 1181 DUPLICATE APPS

In [55]:
print(len(duplicates))

1181


In [56]:
print(duplicates)

[['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up'], ['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device'], ['Google My Business', 'BUSINESS', '4.4', '70991', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 24, 2018', '2.19.0.204537701', '4.4 and up'], ['ZOOM Cloud Meetings', 'BUSINESS', '4.4', '31614', '37M', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 20, 2018', '4.1.28165.0716', '4.0 and up'], ['join.me - Simple Meetings', 'BUSINESS', '4.0', '6989', 'Varies with device', '1,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 16, 2018', '4.3.0.508', '4.4 and up'], ['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31,

The duplicates will be removed based on the number of reviews.
The most recent duplicates have higher number of reviews.

In [57]:
for i in google:
    print(i)
    break

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [77]:
print('Expected length:', len(google[1:]) - 1181)

Expected length: 9660


In [65]:
reviews_max = {}
def conv(x):#function to convert reviews that have letters M and K
    if 'M' in x:
        x= x.strip('M')
        return float(x)*1000000
    elif 'K' in x:
        return float(x)*1000
    return float(x)   
for row in google[1:]:
    name = row[0]#points to the app name
    n_reviews = conv(row[3])#points to the app reviews
    if name in reviews_max and reviews_max[name]<n_reviews:
        reviews_max[name]=n_reviews
    elif name not in reviews_max:
        reviews_max[name]=n_reviews    

In [67]:
print(len(reviews_max))

9660


Created two new lists;
"android_clean" and "already_added"
one to store the new cleaned dataset
and the other to store the names of apps already added to the new dataset.

android_clean list is used to store lists.
already_added is used to store app names.



In [70]:
android_clean = []# to store cleaned dataset
already_added = []# to store app names
for row in google[1:]:
    name = row[0]# name of the app
    n_reviews = conv(row[3])#reviews column
    if n_reviews == reviews_max[name] and name not in already_added:
       android_clean.append(row)
       already_added.append(name)        

In [72]:
print(already_added)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'Sketch - Draw & Paint', 'Pixel Draw - Number Art Coloring Book', 'Paper flowers instructions', 'Smoke Effect Photo Maker - Smoke Editor', 'Infinite Painter', 'Garden Coloring Book', 'Kids Paint Free - Drawing Fun', 'Text on Photo - Fonteee', 'Name Art Photo Editor - Focus n Filters', 'Tattoo Name On My Photo Editor', 'Mandala Coloring Book', '3D Color Pixel by Number - Sandbox Art Coloring', 'Learn To Draw Kawaii Characters', 'Photo Designer - Write your name with shapes', '350 Diy Room Decor Ideas', 'FlipaClip - Cartoon animation', 'ibis Paint X', 'Logo Maker - Small Business', "Boys Photo Editor - Six Pack & Men's Suit", 'Superheroes Wallpapers | 4K Backgrounds', 'HD Mickey Minnie Wallpapers', 'Harley Quinn wallpapers HD', 'Colorfit - Drawing & Coloring', 'Animated Photo Editor', 'Pencil Sketch Drawing', 'Easy Realistic Drawing Tutorial', 'Pink Silver Bow Keyboard Theme', 'Art Dr

Confirmation that final length = expected length; which is 9660

In [78]:
print(len(android_clean))

9660


REMOVING NON-ENGLISH APPS: PART 1

*there are no duplicates in the apple store dataset*

In [82]:
print(android_clean[4412][0])

中国語 AQリスニング


English text has a corresponding number associated with it.
This can be retrieved by using the function **ord()

In [84]:
print(ord('a'))
print(ord('A'))
print(ord('中'))

97
65
20013


To detect foreign characters, create a function that checks if any of the characters is greater than 127; commonly used characters in English are within 0-127 range

In [91]:
def check_foreign(strr):#function checks if there is a foreign character in input string
  for i in strr:
    if ord(i)>127:
        return False
  return True  

Testing the function.

In [93]:
print(check_foreign('Instagram'))
print(check_foreign('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check_foreign('Docs To Go™ Free Office Suite'))
print(check_foreign('Instachat 😜'))

True
False
False
False


Removing Non-English Apps: PART 2

we saw that the function couldn't correctly identify certain English app names like 

'Docs To Go™ Free Office Suite' and 'Instachat 😜'

This is because emojis and characters like ™ fall outside the ASCII range and have corresponding numbers over 127.

In [95]:
print(ord('😜'))
print(ord('™'))

128540
8482


If we're going to use the function we've created, we'll lose useful data since many English apps will be incorrectly labeled as non-English.

To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range.

In [99]:
def check_foreign(strr):#correcting the function to check if more than 3 characters fall outside the Ascii range
    count = 0#variable to count number of outliers; chars that fall outside the Ascii range
    for i in strr:
        if ord(i) > 127:
           count+=1
        if count==3:
            return False
    return True   

In [100]:
print(check_foreign('Docs To Go™ Free Office Suite'))
print(check_foreign('Instachat 😜'))
print(check_foreign('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Applying the function to the Dataset; android_clean

In [103]:
english_apps=[]
for row in android_clean:
    if check_foreign(row[0]):
        english_apps.append(row)

In [104]:
print(len(english_apps))

9598


English Apps: 9598

Isolating the Free Apps

In [119]:
print(google[0])
print(apple[9])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['324684580', 'Spotify Music', '132510720', 'USD', '0.0', '878563', '8253', '4.5', '4.5', '8.4.3', '12+', 'Music', '37', '5', '18', '1']


Index 6 is used to determine whether app is free or not; for android apps. Declared with 'Free'
Index 4 is for apple store. Declared with '0.0'

In [123]:
free_google_apps = []
free_apple_apps = []
for row in android_clean:
    status = row[6]#points to 'Type' column
    if status == 'Free':
        free_google_apps.append(row)
for row in apple[1:]:
    price = row[4]#points to 'Price' column
    if price == '0.0':
        free_apple_apps.append(row)
        

In [129]:
print('There are '+str(len(free_google_apps))+' free android apps')
print('There are '+str(len(free_apple_apps))+' free apple apps')

There are 8904 free android apps
There are 4056 free apple apps


In [133]:
rem_apple = len(apple[1:])-len(free_apple_apps)
rem_android = len(android_clean)-len(free_google_apps) 

print('Apple: {}'.format(rem_apple))
print('Android: {}'.format(rem_android))

Apple: 3141
Android: 756


Most Common Apps by GENRE: Part 1

In [139]:
print(apple[0])
print("\n")
print(google[0])

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


Our goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue.

To achieve this we need to find app profiles that are successful in both stores; Google Play and the App Store.

We will begin by determining the most common genre in each market

In [141]:
print(apple[0:2])
print("\n")
print(google[0:2])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic'], ['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']]


[['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'], ['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']]


index -5 or 11 on apple list and index -4 or 9 on google list will be used to generate the frequency table.

**Most common Apps by Genre: Part 2

*Method 1

In [155]:
freq_apple={}
freq_android={}
for row in free_apple_apps:
    genre = row[-5]#points to the 'prime_genre' column
    if genre in freq_apple:
        freq_apple[genre]+=1
    elif genre not in freq_apple:
        freq_apple[genre]=1
for row in free_google_apps:
    genre = row[-4]#points to the 'prime_genre' column
    if genre in freq_android:
        freq_android[genre]+=1
    elif genre not in freq_android:
        freq_android[genre]=1
              

In [156]:
print(freq_apple)
print(freq_android)

{'Social Networking': 143, 'Photo & Video': 167, 'Games': 2257, 'Music': 67, 'Reference': 20, 'Health & Fitness': 76, 'Weather': 31, 'Utilities': 109, 'Travel': 56, 'Shopping': 121, 'News': 58, 'Navigation': 20, 'Lifestyle': 94, 'Entertainment': 334, 'Food & Drink': 43, 'Sports': 79, 'Book': 66, 'Finance': 84, 'Education': 132, 'Productivity': 62, 'Business': 20, 'Catalogs': 9, 'Medical': 8}
{'Art & Design': 54, 'Art & Design;Creativity': 6, 'Auto & Vehicles': 82, 'Beauty': 53, 'Books & Reference': 194, 'Business': 408, 'Comics': 55, 'Comics;Creativity': 1, 'Communication': 288, 'Dating': 165, 'Education': 480, 'Education;Creativity': 4, 'Education;Education': 31, 'Education;Pretend Play': 5, 'Education;Brain Games': 3, 'Entertainment': 542, 'Entertainment;Brain Games': 7, 'Entertainment;Creativity': 3, 'Entertainment;Music & Video': 15, 'Events': 63, 'Finance': 328, 'Food & Drink': 110, 'Health & Fitness': 273, 'House & Home': 73, 'Libraries & Demo': 83, 'Lifestyle': 349, 'Lifestyle;P

Method 2

We'll build two functions we can use to analyze the frequency tables:

--One function to generate frequency tables that show percentages

--Another function we can use to display the percentages in a descending order

In [148]:
def freq_table(dataset,index,header=False):
    freq_table={}
    if header:
        dataset=dataset[1:]
    for row in dataset:
        cat = row[index]# category of focus
        if cat in freq_table:
            freq_table[cat]+=1
        elif cat not in freq_table:
            freq_table[cat]=1
    for cats in freq_table:#converting frequencies to percentages   
       freq_table[cats]=(freq_table[cats]/len(dataset))*100 
    return freq_table 
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [158]:
display_table(free_apple_apps,11)
print('\n')
display_table(free_google_apps,9)
display_table(free_google_apps,1)

Games : 55.64595660749507
Entertainment : 8.234714003944774
Photo & Video : 4.117357001972387
Social Networking : 3.5256410256410255
Education : 3.2544378698224854
Shopping : 2.983234714003945
Utilities : 2.687376725838264
Lifestyle : 2.3175542406311638
Finance : 2.0710059171597637
Sports : 1.947731755424063
Health & Fitness : 1.8737672583826428
Music : 1.6518737672583828
Book : 1.6272189349112427
Productivity : 1.5285996055226825
News : 1.4299802761341223
Travel : 1.3806706114398422
Food & Drink : 1.0601577909270217
Weather : 0.7642998027613412
Reference : 0.4930966469428008
Navigation : 0.4930966469428008
Business : 0.4930966469428008
Catalogs : 0.22189349112426035
Medical : 0.19723865877712032


Tools : 8.423180592991914
Entertainment : 6.087151841868823
Education : 5.3908355795148255
Business : 4.5822102425876015
Lifestyle : 3.919586702605571
Productivity : 3.8858939802336026
Finance : 3.6837376460017968
Medical : 3.515274034141959
Sports : 3.447888589398023
Personalization : 3.313

**Most common Apps by Genre: Part 3

Majority of the apps in the Apps Store are Games.
But this obviously does not imply that they have the majority of users.
While there is an almost even distribution across all genres in the Google Play Store.

**Most Popular Apps by Genre on the App Store.

*One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre.

 *For the Google Play data set, we can find this information in the **Installs** column, but this information is missing for the App Store data set.
 As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the **rating_count_tot** app.

**Let's start with calculating the average number of user ratings per app genre on the App Store. To do that, we'll need to do the following:

-- Isolate the apps of each genre
-- Add up the user ratings for the apps of that genre
-- Divide the sum by the number of apps belonging to that genre (not by the total number of apps)

In [160]:
app_store=freq_table(free_apple_apps,11)#frequency of unique app genres on App Store
print(app_store)

{'Social Networking': 3.5256410256410255, 'Photo & Video': 4.117357001972387, 'Games': 55.64595660749507, 'Music': 1.6518737672583828, 'Reference': 0.4930966469428008, 'Health & Fitness': 1.8737672583826428, 'Weather': 0.7642998027613412, 'Utilities': 2.687376725838264, 'Travel': 1.3806706114398422, 'Shopping': 2.983234714003945, 'News': 1.4299802761341223, 'Navigation': 0.4930966469428008, 'Lifestyle': 2.3175542406311638, 'Entertainment': 8.234714003944774, 'Food & Drink': 1.0601577909270217, 'Sports': 1.947731755424063, 'Book': 1.6272189349112427, 'Finance': 2.0710059171597637, 'Education': 3.2544378698224854, 'Productivity': 1.5285996055226825, 'Business': 0.4930966469428008, 'Catalogs': 0.22189349112426035, 'Medical': 0.19723865877712032}


Average ratings of genre in the App Store

In [193]:

genre_ratings = {}  # Dictionary to store genre ratings

for row in free_apple_apps:
    genre_app = row[11]  # Save the app genre
    ratings = float(row[5])  # Save the number of ratings of the app
    
    if genre_app in genre_ratings:
        genre_ratings[genre_app]['total'] += ratings
        genre_ratings[genre_app]['count'] += 1
    else:
        genre_ratings[genre_app] = {'total': ratings, 'count': 1}

for genre, data in genre_ratings.items():
    avg = data['total'] / data['count']
    print("{} average rating is {:.2f}".format(genre, avg))
    

Social Networking average rating is 53078.20
Photo & Video average rating is 27249.89
Games average rating is 18924.69
Music average rating is 56482.03
Reference average rating is 67447.90
Health & Fitness average rating is 19952.32
Weather average rating is 47220.94
Utilities average rating is 14010.10
Travel average rating is 20216.02
Shopping average rating is 18746.68
News average rating is 15892.72
Navigation average rating is 25972.05
Lifestyle average rating is 8978.31
Entertainment average rating is 10822.96
Food & Drink average rating is 20179.09
Sports average rating is 20128.97
Book average rating is 8498.33
Finance average rating is 13522.26
Education average rating is 6266.33
Productivity average rating is 19053.89
Business average rating is 6367.80
Catalogs average rating is 1779.56
Medical average rating is 459.75


Average rating of category in Google Play Store

In [ ]:
genre_ratings = {}  # Dictionary to store genre ratings

for row in free_ap_apps:
    genre_app = row[11]  # Save the app genre
    ratings = float(row[5])  # Save the number of ratings of the app
    
    if genre_app in genre_ratings:
        genre_ratings[genre_app]['total'] += ratings
        genre_ratings[genre_app]['count'] += 1
    else:
        genre_ratings[genre_app] = {'total': ratings, 'count': 1}

for genre, data in genre_ratings.items():
    avg = data['total'] / data['count']
    print("{} average rating is {:.2f}".format(genre, avg))